internal/pkg/api/handleUpload.go

```diff
@@ -0,0 +1,388 @@
+
+package api
+
+import (
+	"context"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"hash"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/elastic/fleet-server/v7/internal/pkg/bulk"
+	"github.com/elastic/fleet-server/v7/internal/pkg/cache"
+	"github.com/elastic/fleet-server/v7/internal/pkg/config"
+	"github.com/elastic/fleet-server/v7/internal/pkg/limit"
+	"github.com/elastic/fleet-server/v7/internal/pkg/logger"
+	"github.com/elastic/fleet-server/v7/internal/pkg/upload"
+	"github.com/elastic/fleet-server/v7/internal/pkg/upload/cbor"
+	"github.com/elastic/go-elasticsearch/v7"
+	"github.com/julienschmidt/httprouter"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+// the only valid values of upload status according to storage spec
+type UploadStatus string
+
+const (
+	UploadAwaiting UploadStatus = "AWAITING_UPLOAD"
+	UploadProgress UploadStatus = "UPLOADING"
+	UploadDone     UploadStatus = "READY"
+	UploadFail     UploadStatus = "UPLOAD_ERROR"
+	UploadDel      UploadStatus = "DELETED"
+)
+
+const (
+	// TODO: move to a config
+	maxParallelUploadOperations = 3
+	maxParallelChunks           = 4
+	maxFileSize                 = 104857600 // 100 MiB
+
+)
+
+func (rt Router) handleUploadStart(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	start := time.Now()
+
+	reqID := r.Header.Get(logger.HeaderRequestID)
+
+	zlog := log.With().
+		Str(ECSHTTPRequestID, reqID).
+		Logger()
+
+	// authentication occurs inside here
+	// to check that key agent ID matches the ID in the body payload yet-to-be unmarshalled
+	err := rt.ut.handleUploadStart(&zlog, w, r)
+
+	if err != nil {
+		cntUpload.IncError(err)
+		resp := NewHTTPErrResp(err)
+
+		// Log this as warn for visibility that limit has been reached.
+		// This allows customers to tune the configuration on detection of threshold.
+		if errors.Is(err, limit.ErrMaxLimit) || errors.Is(err, upload.ErrMaxConcurrentUploads) {
+			resp.Level = zerolog.WarnLevel
+		}
+
+		zlog.WithLevel(resp.Level).
+			Err(err).
+			Int(ECSHTTPResponseCode, resp.StatusCode).
+			Int64(ECSEventDuration, time.Since(start).Nanoseconds()).
+			Msg("fail upload initiation")
+
+		if err := resp.Write(w); err != nil {
+			zlog.Error().Err(err).Msg("fail writing error response")
+		}
+	}
+}
+
+func (rt Router) handleUploadChunk(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	start := time.Now()
+
+	id := ps.ByName("id")
+	chunkID := ps.ByName("num")
+
+	reqID := r.Header.Get(logger.HeaderRequestID)
+
+	zlog := log.With().
+		Str(LogAgentID, id).
+		Str(ECSHTTPRequestID, reqID).
+		Logger()
+
+	// simpler authentication check, since chunk checksum must
+	// ultimately match the initial hash provided with the stricter key check
+	if _, err := authAPIKey(r, rt.bulker, rt.ut.cache); err != nil {
+		cntUpload.IncError(err)
+		resp := NewHTTPErrResp(err)
+		if err := resp.Write(w); err != nil {
```

Comment:
Typically you write the http response at the top of the stack.  Not buried down.  Better to return an error and have one block of code that writes out the  http response.